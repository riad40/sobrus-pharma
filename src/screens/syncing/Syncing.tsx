import React, { useEffect, useState } from 'react'
import { View, Text, SafeAreaView, ScrollView, Image } from 'react-native'

import syncingStyles from './syncing.styels'
import { NavBar, ProgressBar } from '../../components'

import { useNavigation, NavigationProp } from '@react-navigation/native'
import RootStackParamsList from '../../navigations/RootStackParamsList'

import { getProducts } from '../../controllers/ProductsController'

import realm from '../../configs/realm'

import { Product } from '../../@types'

console.log('Syncing screen loaded')

const Syncing = (): JSX.Element => {
    const navigation = useNavigation<NavigationProp<RootStackParamsList>>()

    const [progress, setProgress] = useState(0)

    const syncData = async () => {
        try {
            const products: Product[] = (await getProducts()) || []

            realm.write(() => {
                products.forEach(product => {
                    const existingProduct = realm.objectForPrimaryKey('Product', product.id)

                    if (existingProduct) return

                    realm.create('Product', {
                        id: product.id,
                        name: product.name,
                        codeBar: product.codeBar.toString(),
                        status: 'known'
                    })
                })
            })

            setProgress(100)

            setTimeout(() => {
                navigation.navigate('BottomTabs')
            }, 1000)
        } catch (error) {
            console.log('syncing error', error)
        }
    }

    useEffect(() => {
        syncData()
    }, [])

    return (
        <SafeAreaView testID="SyncingScreen">
            <ScrollView style={syncingStyles.screenContainer}>
                <NavBar />

                <View style={syncingStyles.centeredImageContainer}>
                    <Image source={require('../../assets/images/syncing.png')} style={syncingStyles.centeredImage} />
                </View>

                <ProgressBar progress={progress} />

                <View style={syncingStyles.bottomContent}>
                    <Text style={syncingStyles.boldText}>Synchronisation en cours …</Text>
                    <Text style={syncingStyles.lightText}>
                        Nous sommes entrain de mettre à jour votre base de données, Veuillez patienter quelques instants
                    </Text>
                </View>
            </ScrollView>
        </SafeAreaView>
    )
}

export default Syncing
