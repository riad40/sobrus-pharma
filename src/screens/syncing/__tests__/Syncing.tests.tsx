import React from 'react'
import Syncing from '../Syncing'
import { render } from '@testing-library/react-native'
import { NavigationContainer } from '@react-navigation/native'

describe('Syncing Screen', () => {
    jest.useFakeTimers()

    const SyncingScreen = (): JSX.Element => {
        return (
            <NavigationContainer>
                <Syncing />
            </NavigationContainer>
        )
    }

    // test all the screen renders
    it('should render correctly', () => {
        const screen = render(<SyncingScreen />)
        expect(screen).toMatchSnapshot()
    })
})
