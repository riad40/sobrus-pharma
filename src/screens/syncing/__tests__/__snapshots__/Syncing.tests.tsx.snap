// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Syncing Screen should render correctly 1`] = `
<RCTSafeAreaView
  testID="SyncingScreen"
>
  <RCTScrollView
    style={
      {
        "backgroundColor": "#fff",
        "height": "100%",
        "padding": 20,
        "width": "100%",
      }
    }
  >
    <View>
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginVertical": 30,
          }
        }
      >
        <Image
          source={
            {
              "testUri": "../../../src/assets/images/pharmalogo.png",
            }
          }
          style={{}}
        />
        <Image
          source={
            {
              "testUri": "../../../src/assets/images/doctorHeadset.png",
            }
          }
          style={{}}
        />
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "justifyContent": "center",
          }
        }
      >
        <Image
          source={
            {
              "testUri": "../../../src/assets/images/syncing.png",
            }
          }
          style={
            {
              "height": 667,
              "resizeMode": "contain",
              "width": 300,
            }
          }
        />
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#E0FAFA",
              "borderColor": "#00C1B3",
              "borderRadius": 40,
              "borderWidth": 1,
              "flexDirection": "row",
              "height": 40,
              "justifyContent": "center",
              "width": "50%",
            }
          }
        >
          <Image
            collapsable={false}
            source={
              {
                "testUri": "../../../src/assets/images/refresh.png",
              }
            }
            style={{}}
          />
          <Text
            collapsable={false}
            style={
              {
                "color": "#00C1B3",
                "fontFamily": "Poppins-Regular",
                "fontSize": 16,
                "marginTop": 2,
                "paddingLeft": 5,
                "width": "60%",
              }
            }
          >
            Syncing 
            63
            %
          </Text>
          <View
            collapsable={false}
            style={
              {
                "backgroundColor": "#00C1B3",
                "borderBottomLeftRadius": 40,
                "borderTopLeftRadius": 40,
                "height": "100%",
                "left": 0,
                "opacity": 0.2,
                "position": "absolute",
                "width": "0%",
                "zIndex": 1,
              }
            }
          />
        </View>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "justifyContent": "center",
            "marginTop": 20,
          }
        }
      >
        <Text
          style={
            {
              "color": "#000",
              "fontFamily": "Poppins-SemiBold",
              "fontSize": 16,
              "marginBottom": 10,
            }
          }
        >
          Synchronisation en cours …
        </Text>
        <Text
          style={
            {
              "color": "#000",
              "fontFamily": "Poppins-Regular",
              "fontSize": 13,
              "marginBottom": 20,
            }
          }
        >
          Nous sommes entrain de mettre à jour votre base de données, Veuillez patienter quelques instants
        </Text>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
